Step3 - 3단계 - 로또(자동) 코드리뷰요청드립니다.

aksgudwns opened the pull request:
정성스럽게 리뷰해주셔서 감사드리며, 3단계과제 리뷰요청 드립니다.
과제 작업중 궁금한점이 있어서 문의사항 남깁니다..

1. 도메인 기준을 잘못나눴는지 궁금합니다.
전반적으로 일급콜렉션을 너무 많이 사용한 것 같아 클래스명이 길어지고 장황해지고 가독성이 떨어지는 느낌이 있는 것 같습니다..
이대로 괜찮을지 구조를 다시 바꿔서 생각해야할지 문의드립니다..

2. 코드컨벤션을 지키기 위한 스트림사용 문의
프리코스 과제 때 스트림 쓰지말라고 하였던 것 같아서 문의드립니다..
Lotto.java isMatch메소드, getMatchNumber메소드, LottoMatchNumberMap.java getMatchLottoNumber메소드
indent가 두번되어 stream으로 바꾸고 싶은데 바꾸어도 될까요..? stream사용 외에 더 좋은 방법이 있을까요..?
Arrays.stream(winningNumbers).anyMatch(winningNumber-> winningNumber==number);
Arrays.stream(numbers).filter(number -> isMatch(number)).count();
lottoMatchNumberMap.keySet().stream().filter(lotto -> lottoMatchNumberMap.get(lotto) == matchNumber).count();

4. LottoResult.java winningPriceMap을 jdk 11버전의 Map.of(key,value)를 사용하여 선언과 동시에 초기화 하고 싶은데.. 
저희 과제제출기준은 jdk 8버전이여서 한번에 map을 선언할 수 있는 다른 좋은 방법이 있을까요? 
map사용말고, enum으로 빼는게 맞을까요..

5. InputView, OutputView 메시지구현을 controller, domain, view 나누는 기준이 궁금합니다.
저는 InputView는 공통으로 잡고 컨트롤러에서 사용자 메시지를 주입하였습니다.
그런데 사용자 메시지 출력은 View에서만 이루어지는게 맞는 것 같기도하여서, 클래스를 다 만들어서 각각 메시지를 넣어두도록 수정해야 할지 문의드립니다.
OutputView 메시지 출력도 공통으로 잡고, domain객체의 toString을 통해서 작업하였습니다.
처음에는 domain객체 데이터만 전달하고, OutputView에서 메시지 출력을 구현하려 했지만, 
객체지향 원칙에서 getter, setter를 지양하라는 룰이 있어서 domain에서 toString을 통해 메시지를 구현하였습니다.
domain객체를 MVC패턴에서의 model이라 생각하고 데이터 전달의 의미로 getter를 허용해주고, 메시지 출력은 OutputView클래스 각각에서 하는게 더 올바른 방법인지 문의드립니다.

감사합니다.

aksgudwns committed:
fix - setTest 수정사항 반영

assertThat을 통한 체크로 수정

aksgudwns committed:
doc - 문자열 덧셈 계산기 요구사항 추가

문자열 덧셈 계산기 요구사항 추가

aksgudwns committed:
feat - 문자열계산기 빈문자null체크, 숫자 유효성체크 기능 작성

- 빈문자null체크 기능 작성
- 입력된 문자열이 숫자인지 체크하는 유틸로직 작성

aksgudwns committed:
feat - 구현기능 테스트, splitor만 입력했을 때 오류 추가

- 기본기능 테스트
- splitor만 입력했을 경우 오류나도록 추가

aksgudwns committed:
feat - custom delimiter 기능 구현

- custom delimiter로 구분하여 합을 구할 수 있는 기능 추가
- custom delimiter와 default delimiter를 혼합하였을 경우 테스트케이스 추가

aksgudwns committed:
doc - 로또 요구사항 추가

로또 요구사항 추가

aksgudwns committed:
feat - 로또생성, 일치하는 숫자 찾는 기능 추가

일치하는 숫자 찾는 기능 추가

aksgudwns committed:
feat - lotteries 기능구현

- 구매금액 입력시 구매가능한 개수만큼 로또 생성
- 당첨번호가 주어졌을 때 로또당 당첨횟수가 매핑되어 있는 map구현체 생성

aksgudwns committed:
feat - LottoMatchNumberMap 구현

- 로또당 당첨횟수가 매핑되어 있는 맵 구현
- n개 일치하는 로또가 몇개있는지 구하는 기능 구현

aksgudwns committed:
feat - LottoResult 구현

- LottoResult 결과 메시지 구현
- 상금 금액 계산, 수익률계산

aksgudwns committed:
feat - 화면로직추가

- controller, service추가

aksgudwns committed:
Refactor - 코드리뷰 반영

- 자바코드컨벤션 한 줄짜리에서도 중괄호 사용하도록 수정
- Pattern.compile("//(.)\n(.*)") 비즈니스로직상 메서드 호출할때마다 인스턴스 생성될 필요 없으므로 상수로 변경
- StringAddCalculator splitter 상수화
- StringAddCalculatorUtil 제거 Util 클래스는 좀 더 보수적으로 생각
- 테스트메서드 같은 목적을 가진 메서드들은 ParameterizedTest로 통합
- RuntimeException IllegalArgumentException과 메시지 추가로 오류내용 구체화

aksgudwns committed:
Merge branch 'step2' into step3

aksgudwns committed:
코드리뷰 소스와 충돌해결

aksgudwns committed:
Refactor - LottoResultView추가, InputView, 가독성을 위한 메소드, 클래스 네이밍변경

- 사용자에게 보여주는 출력을 담당하는 LottoResultView 구현
- 요구사항에 따라 Reader -> View로 네이밍 변경
- if 한줄도 {}추가 코드컨벤션 따라서 수정
- 테스트 메소드 이름 룰 통일 (메소드명_테스트하고자하는내용_테스트)
- Parameterized할 수 있는 테스트들 하나의 메소드로 통합

aksgudwns committed:
refactor - InputView LottoInputView로 통일, 유효성체크 추가

- 공통기능을 묶어서 LottoInputView로 통일
- 유효성체크 추가
구매금액 1000원이상
당첨번호 중복불가
당첨번호 1~45
총 수익률1 분기에 따른 문구노출 결정

dhmin5693 commented:
리뷰에 앞서 남겨주신 질문에 대한 답변 먼저 남깁니다.

1. 도메인 기준을 잘못 남겼는가?
    - 더 나은게 있을 뿐, 잘못이란건 없다는 방향으로 접근해주시는게 더 좋다는 의견 드립니다. 또한 지금 구분되어 있는 도메인도 나쁘다고는 생각되지 않습니다.

2. 스트림 사용 문의
    - 스트림은 자바8이 남긴 가장 훌륭한 업적 중 하나입니다. 프리코스에선 사용하지 말라는 말이 있었으니 쓰지 않는게 맞지만 이번 미션에선 그런 말이 없었으니 사용하셔도 무방합니다.
    - 그거랑 별개로 별도의 라이브러리(롬복 등)은 학습을 위해 사용하지 않는 것을 권장드립니다.

3. java 8에서의 Map 초기화
    - [Double brace](https://www.baeldung.com/java-double-brace-initialization) 라는 방법이 있긴 한데, 권장되지는 않습니다. 저는 리뷰할 때 이 문법이 보이면 쓰지 않는 방향으로 가이드를 드리고 있어요.
    - 이거 말고는 Guava 같은 외부 라이브러리를 사용하는 방법이 있는데... 미션 내에서는 불가능하다고 보는게 맞겠네요.

4. 메시지 구현 기준
    - 스프링 기준으로 생각해보면, controller에서 외부 요청을 받고 서비스 로직(도메인)을 거친 뒤 View(웹프론트)로 반환합니다.
    - 최대한 비슷한 방법으로 한다고 생각해보시면 어떨까요?

dhmin5693 commented on src/main/java/lotto/domain/LottoMatchNumberMap.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Map;
+
+public class LottoMatchNumberMap {
```
클래스명이 장황해진다는 부분이 이쪽이실까요?
컬렉션은 이름으로 쓰지 않는게 좋기는 합니다.
https://tecoble.techcourse.co.kr/post/2020-04-24-variable_naming/

개인적인 의견으론 `LottoNumberMatcher` 정도로만 해도 괜찮지 않을까요?

aksgudwns replied:
넵.. 앞으로는 컬렉션을 이름으로 사용하지 않겠습니다..

dhmin5693 commented on src/main/java/lotto/domain/LottoMatchNumberMap.java:
```diff
@@ -0,0 +1,23 @@
+
+    public int getMatchLottoNumber(int matchNumber) {
+        int matchLottoNumber = 0;
+        for(Lotto lotto : lottoMatchNumberMap.keySet()) {
+            if(lottoMatchNumberMap.get(lotto) == matchNumber) {
+                matchLottoNumber++;
+            }
+        }
```
이 부분 때문에 stream을 사용하고 싶다고 하신 것 같습니다.
stream을 사용하지 않더라도 if절부터 메소드로 빼내면 충분히 indent 2단계 요구사항을 충족할 수 있습니다.

aksgudwns replied:
요구사항 충족을 위해 stream사용하겠습니다!

dhmin5693 commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,43 @@
+    public LottoResult (Lotteries lotteries, int[] winningNumbers, int buyAmount) {
+        this.lottoMatchNumberMap = new LottoMatchNumberMap(lotteries, winningNumbers);
+        this.buyAmount = buyAmount;
+        this.winningPriceMap =  new HashMap<>();
+        this.winningPriceMap.put(3, 5000);
+        this.winningPriceMap.put(4, 50000);
+        this.winningPriceMap.put(5, 1500000);
+        this.winningPriceMap.put(6, 2000000000);
```
`winningPriceMap` 은 더 이상 수정이 필요하지 않아보입니다.
`Collections.UnmodifiableMap` 같은걸 사용해보는건 어떤가요?

aksgudwns replied:
Collections.UnmodifiableMap 변경하겠습니다!

dhmin5693 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,31 @@
+
+public class LottoController {
+
+    LottoService lottoService = new LottoService();
+    LottoInputView lottoInputView = new LottoInputView();
+    LottoResultView lottoResultView = new LottoResultView();
```
접근 제어자가 private이어도 괜찮을 것으로 보이는데 default인 이유가 있을까요?
final까지 붙여 불변으로 만들어도 좋을거라고 생각됩니다.

aksgudwns replied:
말씀들으니, default 생성자보다 private생성자가 더 어울리는 것으로 보입니다 
불변객체로도 수정하겠습니다

dhmin5693 commented on src/main/java/lotto/domain/Lotteries.java:
```diff
@@ -0,0 +1,40 @@
+    public Lotteries(int buyAmount) {
+        int lottoCount = buyAmount/LOTTO_AMOUNT ;
+        this.lotteries = new ArrayList<>();
+        while(lottoCount-->0) {
```
```suggestion
        while (lottoCount-- > 0) {
```
읽는 입장에서는 실수하기 좋은 문법이라고 생각됩니다.
컨벤션을 맞춰 제어문, 반복문, 뒤에선 한 칸 띄워주시고 구분하기 좋게 만드는건 어떤가요?
같은 로직을 유지한다면 가급적 for가 가독성 면에서 더 나아보입니다.

aksgudwns replied:
컨벤션 유지하면서 가독성을 위해 for문으로 수정하겠습니다

dhmin5693 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+    public static final int LOTTO_MIN_NUMBER = 1;
+    public static final int LOTTO_MAX_NUMBER = 45;
+    public static final int digit = 6;
+    private int[] winningNumbers = new int[digit];
```
배열을 사용해야 하는 이유가 있을까요?
미션 내에서 꽤 많이 보이는데요, 특별한 이유가 없다면 배열을 한 번 가공하여 사용하는 Collection 으로 대체하시는걸 추천드립니다.
아시다시피 편의성 부분에서 꽤 많은 차이가 납니다.

aksgudwns replied:
List로 전부 수정하였습니다!

dhmin5693 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+    private int[] winningNumbers = new int[digit];
+
+    public Lotto() {
+        List<Integer> numbers = new ArrayList();
+        for(int i=LOTTO_MIN_NUMBER; i<=LOTTO_MAX_NUMBER; i++) {
+            numbers.add(i);
+        }
+        Collections.shuffle(numbers);
+        for(int i=0; i< winningNumbers.length; i++) {
+            winningNumbers[i] = numbers.get(i);
+        }
+        Arrays.sort(winningNumbers);
```
혹시 멱등성 문제에 대해 들어보신 적 있으실까요?
설명은 아래 블로그로 대체합니다.
https://jojoldu.tistory.com/451

멱등성이 지켜지지 않는 코드는 개발자가 결과를 정확하게 제어를 할 수 없다는 문제를 가지고 있습니다.
예를 들어보자면, `Lotto`를 테스트하려고 새 로또 인스턴스를 만듭니다.
`1, 2, 3, 4, 5, 6` 번호를 가진 로또를 의도적으로 생성할 수 있을까요?
운이 좋으면 되긴 할건데 시도를 꽤 많이 해야될거에요.

책임을 다른 곳으로 옮기는 것으로 문제를 해결할 수 있습니다.
`Lotto`는 반드시 6개의 숫자를 가져야만 유효합니다.
빈 생성자를 제거하고 6개의 숫자를 매개변수로 주입받는 생성자만 유지해주세요.

그리고 생성하는 역할은 별도의 클래스를 만들어서 책임을 이관하는건 어떤가요?
`LottoCreator` 같은 인터페이스를 만들어서 사용하면 각 객체 사이를 느슨하게 결합할 수 있습니다.

aksgudwns replied:
앞으로 코딩할때 멱등성에도 신경쓰겠습니다.
LottoCreator 인터페이스를 만들고 AutoLottoCreator클래스를 구현해보았습니다.

dhmin5693 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+    public int getMatchNumber(int[] numbers) {
+        int matchNumber = 0;
+        for(int number : numbers) {
+            if(isMatch(number)) {
+                matchNumber++;
+            }
```
여기도 indent 요구조건을 충족하지 못하고 있네요.
수정 부탁드립니다.

aksgudwns replied:
stream사용하여 수정하겠습니다..

dhmin5693 commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,15 @@
+        Lotteries lotteries = lottoController.buyLotto(buyAmount);
+        int[] winningNumbers = lottoController.readWinningNumbers();
+        lottoController.lottoResult(lotteries, winningNumbers, buyAmount);
+    }
```
사람마다 다르겠지만 TDD 미션에서 가장 좋아보이는 아키텍처는 이런 구조라고 생각합니다.

1. main은 controller과 view 패키지에 접근 가능
2. controller는 특정 역할을 수행하고 결과를 반환
3. view 는 입력을 받아 controller에 전달하거나 controller로부터 결과를 받아 출력

이런 논리로 생각해보면 controller에서 입력을 받기보다는 view 에서 해당 역할을 전부 수행하는건 어떤가요?

aksgudwns replied:
말씀 주신 의견 참고해서 구현하도록 노력해보았습니다!
main에서 controller와 view에만 접근 가능하도록 작업해보니, 자연스레 controller 메소드에서 Dto위주로 in/out이 이루어지게 된 것 같습니다.

dhmin5693 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public class LottoService {
```
이번 미션에서는 스프링의 레이어드 아키텍처보다는 객체지향 기반의 코딩 스타일을 권장드립니다.
이 클래스의 내용은 도메인 등으로 충분히 옮겨갈 수 있을 것 같아요.
서비스는 삭제 부탁드리겠습니다.

aksgudwns replied:
말씀주신대로 서비스를 삭제하고 각각의 도메인에서 로직처리하겠습니다!

dhmin5693 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,75 @@
+        return buyAmount;
+    }
+
+    public int[] readWinningNumbers(String userInput) {
+        String[] userInputs = userInput.split(",");
+        winningNumberSizeCheck(userInputs);
+        return getWinningNumbers(userInputs);
+    }
```
당첨 번호를 별개의 클래스에 책임을 부여해 보는건 어떨까요?
winningNumberSizeCheck 또한 그 클래스에서 수행할 수 있습니다.

이런 방식으로 접근하면 이 클래스의 메소드를 하나씩 제거할 수 있을거에요.

aksgudwns replied:
WinningNumbers클래스를 새로 구현해보았습니다!
Lotto 클래스를 상속받아서 만들어보았습니다.

dhmin5693 commented on src/main/java/lotto/view/LottoInputView.java:
```diff
@@ -0,0 +1,19 @@
+
+    public String readUserInput(String userMessage){
+        System.out.println(userMessage);
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
```
미션에선 `Scanner` 사용을 안내하긴 했으나 `BufferedReader` 도 유사하게 사용할 수 있습니다.
그리고 이 두 개의 공통점은 한 번만 선언하고 재활용하여 리소스를 아낄 수 있다는 점입니다.
`BufferedReader` 를 상수로 만드는건 어떨까요?

aksgudwns replied:
상수로 빼도록 하겠습니다!

dhmin5693 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+        Arrays.sort(winningNumbers);
+    }
+
+    public Lotto(int[] winningNumbers) {
+        this.winningNumbers = winningNumbers;
+    }
```
여기도 마찬가지로 배열보다는 가급적 `List` 같은 컬렉션을 사용해주세요 🙏

aksgudwns replied:
전부 수정하겠습니다..

dhmin5693 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public boolean isMatch(int number) {
+        for(int i=0; i<winningNumbers.length; i++) {
+            if(winningNumbers[i] == number) {
+                return true;
+            }
+        }
```
여기도 indent 확인해주세요!

aksgudwns replied:
stream사용으로 변경하겠습니다..

dhmin5693 commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Override
+    public String toString() {
```
toString에 너무 많은 책임을 부여한건 아닐까요?
이 메소드는 출력하기에 간편한 `String` 으로 변환한다는 책임만 있으면 될 것 같습니다.
내부에서 계산도 하고 있네요.

만약 view까지 이 많은 내용을 전달하는게 어렵다면 dto로 전달해보는건 어떤가요?
view에서 dto까지는 참조해도 괜찮다고 생각합니다.

aksgudwns replied:
domain에서 getDto하는 메서드를 추가로 작성하여 View단으로 데이터 주고 받게 작업해보았습니다.

dhmin5693 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+
+public class Lotto {
+
+    public static final int LOTTO_MIN_NUMBER = 1;
+    public static final int LOTTO_MAX_NUMBER = 45;
+    public static final int digit = 6;
```
digit은 위 상수 두 개처럼 컨벤션을 맞춰주세요.
LOTTO_MIN_NUMBER나 LOTTO_MAX_NUMBER 는 적절하게 책임을 분배한다면 private으로 선언되어도 괜찮지 않을까요?
이를테면 로또 번호 자체를 하나의 도메인으로 만든 `LottoNumber`은 어떻게 생각하세요?

aksgudwns replied:
LottoNumber클래스를 새로 만들어보았습니다.
컨벤션도 주의하겠습니다

dhmin5693 commented on src/test/java/lotto/LotteriesTest.java:
```diff
@@ -0,0 +1,55 @@
+    int[] winningNumbers = new int[] {1,2,3,4,5,6};
+
+    @Test
+    void getMatchNumber_여섯개_매치_테스트() {
```
테스트 메소드명에 표현을 해도 괜찮지만 junit5부터는 `@DisplayName` 을 활용하는게 조금 더 좋은 방식입니다.
`String`을 매개변수로 받으므로 메소드명에 한글을 사용하는 것보다 제약없이 표현이 가능합니다.

aksgudwns replied:
@Displayname을 사용하여 메소드명을 더 구체화해보았습니다!

dhmin5693 commented on src/test/java/lotto/LotteriesTest.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Test
+    void getMatchNumber_한개_매치_테스트() {
```
테스트 메소드가 길어지면 주석을 달아주어 가독성을 챙겨보는건 어떨까요?
```java
// given
// when
// then
```
3개만 있어도 굉장히 큰 도움이 됩니다.

aksgudwns replied:
주석 추가하였습니다.
테스트방식 이해가 힘들어 보일 부분은 커멘트도 달아두었습니다!

dhmin5693 reviewed (changes requested):
안녕하세요 형준님!

정말 많은 고민을 하셨음이 코드에서 보이는 것 같습니다.

여기서 정말 조금만 바꿔보면 큰 인사이트를 얻을 수 있다는 생각이 들었습니다.
그런 방향으로 의견을 남겨보았어요. 
이번 미션을 통해 형준님께서 많은 것을 얻어가셨으면 합니다.

최대한 자세하게 피드백을 남겨보려고 했는데요, 만약 난해하거나 어렵다면 부담없이 DM 보내주세요 😁
혼자 고민하기보다는 여럿이서 논의하고 방향을 잡아가는 것도 굉장히 좋은 방법이거든요.

그럼 조금만 더 힘내주세요 💪

aksgudwns committed:
refactor - 코드리뷰반영

- 클래스명 컬렉션이름에서 변경
- indent 2줄이상인 부분 stream활용하여 변경
- winninPriceMap Collections.UnmodifiableMap을 통한 불편객체로 선언
- private 접근자 빠진부분 추가
- 가독성 향상을 위한 for문 사용
- array -> List변경
- 멱등성 원칙에 위배되는 기능 제거 및 수정개발
- 서비스레이어 삭제 후 WinningNumbers, LottoNumber도메인 추가
- 테스트로직 @DisplayName 추가, 가독성을 위한 주석 추가

aksgudwns commented:
세세하게 피드백 해주셔서 감사합니다.
과제하면서 궁금했던 부분이 많이 해결되었습니다!

꼼꼼히 봐주신 덕에 TDD, DDD 개념이 막연하게 있었는데, 많이 구체화 되어가고 있습니다.
특히나 원시값을 객체로 표현하는게 처음엔 메모리 낭비라고 생각했는데, 이번 리뷰작업을 통해 생각이 180도 바뀌었습니다.
TDD로 테스트케이스를 구현해놓으니, 리팩토링 부담이 덜한 점도 무척 좋았습니다.

의견주신 내용 제 나름대로 열심히 고민하고 반영해보았습니다!
부족한 부분은 추가 피드백 부탁드리겠습니다..
덕분에 좋은 개발문화 경험하고 있어서 즐겁습니다 감사드립니다!

aksgudwns committed:
refactor - primitive값 상수로 변경

- 6 -> DIGIT 수정

dhmin5693 commented on src/main/java/lotto/domain/AutoLottoCreator.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Collections;
+import java.util.List;
+
+public class AutoLottoCreator implements LottoCreator {
```
💯

dhmin5693 commented on src/main/java/lotto/domain/AutoLottoCreator.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public Lotteries createLotteries(int buyLottoCount) {
+        List<Lotto> lotteries = new ArrayList();
+        for (int i=0; i<buyLottoCount; i++) {
```
```suggestion
        for (int i = 0; i < buyLottoCount; i++) {
```
여기도 한 칸씩만 공백 넣어주세요 :)
컨벤션이기도 하고 가독성도 좋아집니다.

aksgudwns replied:
지난번에 피드백주셔서 신경써서 찾아가지고 다 고쳤다고 생각했는데.. 빠진부분이 있었군요ㅠㅠ 
수정하고 앞으로 주의하겠습니다!

dhmin5693 commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,17 @@
+    private List<LottoNumber> winningNumbers;
+
+    public WinningNumbers(String winningNumbersString) {
+        super(Arrays.asList(winningNumbersString.split(",")).stream()
+                .map(winningNumberString -> new LottoNumber(winningNumberString))
+                .collect(Collectors.toList()));
```
```suggestion
        super(Arrays.stream(winningNumbersString.split(","))
                    .map(LottoNumber::new)
                    .collect(Collectors.toList()));
```
`Arrays.asList` 로 `List`를 생성하고 만들기보다는 `Arrays.stream(arr[])` 을 바로 활용하면 공간 낭비를 줄일 수 있어요.
두 번째 줄은 람다로 변경이 가능합니다.

그리고 가급적이면 파싱의 책임은 분리해주는게 좋습니다.
예상 밖의 문자가 들어올 수 있어서 애초에 만들 때 제대로 된 걸 받아서 만드는걸 추천드려요.

aksgudwns replied:
Arrays.stream(arr[]), 람다 수정반영하겠습니다!
StringSplitter인터페이스를 만들고 파싱의 책임을 인터페이스를 구현한 다른 클래스로 이관해서 작업해보겠습니다. 
다음리뷰때 피드백 부탁드리겠습니다.

감사합니다!

dhmin5693 reviewed (approved):
코드가 너무 깔끔해졌습니다 💯 💯 
너무 많은 요청을 드렸나 싶어 살짝 죄책감도 있었는데 다 씻겨져 나가는 기분이네요.

리뷰를 남겨두긴 했지만 이미 너무 잘 해주셔서 굳이 이번 단계에 포함하지는 않아도 될 것 같습니다.
천천히 확인만 해주시고 다음 단계 진행해주세요 :)
수고하셨습니다 🙇‍♂️
다음 미션에서 뵈어요~
